File: packages/core/src/report.ts
import { computeBrierScore, computeECE, computeReliabilityDiagram } from "./calibration.js";
import type { CalibrationSample, CorrectnessComparator, DecisionStore, ReliabilityBin } from "./types.js";

export interface CalibrationReport {
  n: number;
  ece: number;
  brier: number;
  reliability: ReliabilityBin[];
}

export interface GetCalibrationOptions {
  decisionName: string;
  field: string;
  numBins?: number;
  /** Defaults to strict equality — override for fields needing tolerance-based correctness (e.g. numeric). */
  comparator?: CorrectnessComparator;
}

const defaultComparator: CorrectnessComparator = (predicted, actual) => predicted === actual;

/**
 * Joins logged decisions with their attached outcomes and computes calibration
 * metrics for one field. Records missing an outcome, or whose recorded probability
 * is null (no usable confidence signal), are excluded rather than treated as wrong.
 */
export async function getCalibration(
  store: DecisionStore,
  options: GetCalibrationOptions,
): Promise<CalibrationReport> {
  const comparator = options.comparator ?? defaultComparator;
  const records = await store.getRecords({ decisionName: options.decisionName });

  const samples: CalibrationSample[] = [];
  for (const record of records) {
    const prediction = record.fields[options.field];
    const outcome = record.outcomes[options.field];
    if (!prediction || !outcome || prediction.probability === null) continue;
    samples.push({
      predictedProbability: prediction.probability,
      correct: comparator(prediction.value, outcome.actualValue),
    });
  }

  return {
    n: samples.length,
    ece: computeECE(samples, options.numBins),
    brier: computeBrierScore(samples),
    reliability: computeReliabilityDiagram(samples, options.numBins),
  };
}

File: python/src/jevcal/report.py
from __future__ import annotations

from typing import Callable

from pydantic import BaseModel

from .calibration import compute_brier_score, compute_ece, compute_reliability_diagram
from .store import JsonlStore
from .types import CalibrationSample, ReliabilityBin

CorrectnessComparator = Callable[[object, object], bool]


class CalibrationReport(BaseModel):
    n: int
    ece: float
    brier: float
    reliability: list[ReliabilityBin]


def _default_comparator(predicted: object, actual: object) -> bool:
    return predicted == actual


def get_calibration(
    store: JsonlStore,
    *,
    decision_name: str,
    field: str,
    num_bins: int = 10,
    comparator: CorrectnessComparator | None = None,
) -> CalibrationReport:
    """Joins logged decisions with their attached outcomes and computes
    calibration metrics for one field. Records missing an outcome, or whose
    recorded probability is None (no usable confidence signal), are excluded
    rather than treated as wrong."""
    comparator = comparator or _default_comparator
    records = store.get_records(decision_name=decision_name)

    samples: list[CalibrationSample] = []
    for record in records:
        prediction = record.fields.get(field)
        outcome = record.outcomes.get(field)
        if prediction is None or outcome is None or prediction.probability is None:
            continue
        samples.append(
            CalibrationSample(
                predicted_probability=prediction.probability,
                correct=comparator(prediction.value, outcome.actual_value),
            )
        )

    return CalibrationReport(
        n=len(samples),
        ece=compute_ece(samples, num_bins),
        brier=compute_brier_score(samples),
        reliability=compute_reliability_diagram(samples, num_bins),
    )
